astropy/io/ascii/qdp.py

```diff
@@ -0,0 +1,552 @@
+            terr_cols = err_specs.pop("terr", [])
+        if serr_cols != []:
+            col_string = " ".join([str(val) for val in serr_cols])
+            print(f"READ SERR {col_string}", file=fobj)
+        if terr_cols != []:
+            col_string = " ".join([str(val) for val in terr_cols])
+            print(f"READ TERR {col_string}", file=fobj)
+
+        if 'comments' in table.meta and table.meta['comments'].strip() != "":
+            for line in table.meta['comments'].split("\n"):
+                line = line.strip()
+                if not line.startswith("!"):
+                    line = "!" + line
+                print(line, file=fobj)
+
+        colnames = table.colnames
+        print("!" + " ".join(colnames), file=fobj)
+        for row in table:
+            values = []
+            for val in row:
+                if not np.ma.is_masked(val):
+                    rep = str(val)
+                else:
+                    rep = "NO"
+                values.append(rep)
+            print(" ".join(values), file=fobj)
+
+
+class QDPSplitter(core.DefaultSplitter):
+    """
+    Split on comma for QDP tables
+    """
+    delimiter = ' '
+
+
+class QDPHeader(basic.CommentedHeaderHeader):
+    """
+    Header that uses the :class:`astropy.io.ascii.basic.QDPSplitter`
+    """
+    splitter_class = QDPSplitter
+    comment = "!"
+    write_comment = "!"
+
+
+class QDPData(basic.BasicData):
+    """
+    Data that uses the :class:`astropy.io.ascii.basic.CsvSplitter`
+    """
+    splitter_class = QDPSplitter
+    fill_values = [(core.masked, 'NO')]
+    comment = "!"
+    write_comment = None
+
+
+class QDP(basic.Basic):
+    """QDP table.
+
+    This file format can contain multiple tables, separated by a line full
+    of ``NO``s. Comments are exclamation marks, and missing values are single
+    ``NO`` entries.
+    Headers are just comments, and tables distributed by various missions
+    can differ greatly in their use of conventions. For example, light curves
+    distributed by the Swift-Gehrels mission have an extra space in one header
+    entry that makes the number of  inconsistent with the number of columns
+
+    ..
+                      Extra space
+                          |
+                          v
+       !     MJD       Err (pos)       Err(neg)        Rate            Error
+        53000.123456   2.378e-05     -2.378472e-05     NO             0.212439
+
+
+    """
+    _format_name = 'qdp'
+    _io_registry_format_aliases = ['qdp']
+    _io_registry_can_write = True
+    _io_registry_suffix = '.qdp'
+    _description = 'Quick and Dandy Plotter'
+
+    header_class = QDPHeader
+    data_class = QDPData
+
+    def __init__(self, table_id=0, input_colnames=None):
+        super().__init__()
+        self.table_id = table_id
+        self.input_colnames = input_colnames
+
+    def read(self, qdp_file):
+        return read_table_qdp(qdp_file, table_id=self.table_id,
```

Comment:
Another key point for API consistency is to allow the input to be a filename, a string with the data, a list of strings or a file object. This is normally done using the following:
```
        # Get a list of the lines (rows) in the table
        self.lines = self.inputter.get_lines(table, newline=newline)
```
Then the core table parsing should always expect a list of lines. I see you have support for providing the table data as a string, though this is not obvious from the docstrings.

To go 100% in the style of `io.ascii`, the standalone routines like `read_table_qdp` would not be considered part of the public API and then you would rewrite that function and all the downstream ones to accept only a list of lines.

Another option is to adapt `get_lines_from_file` to support the standard input options.